WordPress/openverse
# Prevent unnecessary post-process results recursion

## Fixes



Fixes #2415 by @sarayourfriend 

## Description


Stops `_post_process_results` recursion once the total available hits have been exhausted. Previously, if queries were unable to fill the page_size with live links, `_post_process_results` would continue to iterate until the ES max page depth was met.

For some reason the `Hit` `__repr__` was suddenly failing for me and causing tests to incorrectly fail (call argument mocks use `repr`, apparently). The fix for this was to ensure the document `_id` is a string. Here is the error I was seeing:

```
self = <[TypeError('sequence item 1: expected str instance, int found') raised in repr()] Hit object at 0x7f34bb326c90>

    def __repr__(self):
        return "<Hit({}): {}>".format(
>           "/".join(
                getattr(self.meta, key) for key in ("index", "id") if key in self.meta
            ),
            super(Hit, self).__repr__(),
        )
E       TypeError: sequence item 1: expected str instance, int found
```

`self.meta.id` was an `int`. Changing the factory to pass a string here (which does match ES's actual response type) fixed the issue.

## Testing Instructions

Make sure you understand how to reproduce the error on `main`, follow the reproduction instructions in the issue.

Once you've done that, check out this branch and try to reproduce the issue. You should not be able to. With the example in the issue, you should see a single call to `post_process_results`.

The unit test attempts to cover the case where additional calls to `post_process_results` are needed to fill the first or last page of a query. It becomes increasingly complex to test the more calls to `post_process_results` we want to assert, but I can try to add more unit tests if we want examples of recursion depth greater than 2 in the unit tests. The unit tests are also fairly complex, if reviewers have any ideas for how to simplify (and especially how to _clarify_ them), please share.

## Checklist


- [N/A] I ran the DAG documentation generator (if applicable).

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
